ratchet-tree.go

```diff
@@ -0,0 +1,479 @@
+	Info      *ParentNodeInfo `tls:"optional"`
+	LeftHash  []byte          `tls:"head=1"`
+	RightHash []byte          `tls:"head=1"`
+}
+
+type LeafNodeInfo struct {
+	PublicKey  HPKEPublicKey
+	Credential Credential
+}
+
+type LeafNodeHashInput struct {
+	HashType uint8
+	Info     *LeafNodeInfo `tls:"optional"`
+}
+
+//// ratchet tree node and helpers
+type OptionalRatchetNode struct {
+	Node *RatchetTreeNode
+	Hash []byte `tls:"head=2"`
+}
+
+func (n *OptionalRatchetNode) setLeafHash(cs CipherSuite) {
+	lhi := LeafNodeHashInput{HashType: 0}
+	if n.Node != nil {
+		p := n.Node.PublicKey
+		c := n.Node.Cred
+		if c == nil {
+			panic(fmt.Errorf("mls.rtn: Leaf node not provisioned with a credentialp"))
+		}
+		lhi.Info = &LeafNodeInfo{
+			PublicKey:  *p,
+			Credential: *c,
+		}
+	}
+
+	h, err := syntax.Marshal(lhi)
+	if err != nil {
+		panic(fmt.Errorf("mls.rtn: Marshal error %v", err))
+	}
+	n.Hash = cs.digest(h)
+
+}
+
+func (n *OptionalRatchetNode) merge(o *RatchetTreeNode) {
+	if n.Node == nil {
+		n.Node = o
+	} else {
+		n.Node.Merge(o)
+	}
+}
+
+func (n *OptionalRatchetNode) setHash(cs CipherSuite, l OptionalRatchetNode, r OptionalRatchetNode) {
+	phi := ParentNodeHashInput{HashType: 1}
+	if n.Node != nil {
+		phi.Info = &ParentNodeInfo{PublicKey: n.Node.PublicKey, UnmergedLeaves: n.Node.UnmergedLeaves}
+	}
+	phi.LeftHash = l.Hash
+	phi.RightHash = r.Hash
+	data, err := syntax.Marshal(phi)
+	if err != nil {
+		panic(fmt.Errorf("mls.rtn: set hash error %v", err))
+	}
+	n.Hash = cs.digest(data)
+}
+
+func (n *OptionalRatchetNode) hasPrivate() bool {
+	if n.Node != nil && n.Node.PrivateKey != nil {
+		return true
+	}
+	return false
+}
+
+//// RatchetTree Node
+type RatchetTreeNode struct {
+	Cred           *Credential `tls:"optional"`
+	PublicKey      *HPKEPublicKey
+	PrivateKey     *HPKEPrivateKey
+	UnmergedLeaves []leafIndex `tls:"head=2"`
+	CipherSuite    CipherSuite
+}
+
+func newRatchetTreeNode(cs CipherSuite, secret []byte) *RatchetTreeNode {
+	if len(secret) > 0 {
+		priv, err := cs.hpke().Derive(secret)
+		if err != nil {
+			panic(fmt.Errorf("hpke private key generation failed: %v", err))
+		}
+
+		return &RatchetTreeNode{
+			Cred:           nil,
+			PrivateKey:     &priv,
+			PublicKey:      &priv.PublicKey,
+			UnmergedLeaves: []leafIndex{},
+			CipherSuite:    cs,
+		}
+	}
+
+	return &RatchetTreeNode{
+		Cred:           nil,
+		PrivateKey:     nil,
+		PublicKey:      nil,
+		UnmergedLeaves: []leafIndex{},
+		CipherSuite:    cs,
+	}
+}
+
+func (n *RatchetTreeNode) Merge(o *RatchetTreeNode) {
+	if o.PublicKey != nil && !o.PublicKey.equals(n.PublicKey) {
+		n.PublicKey = o.PublicKey
+		n.PrivateKey = nil
+	}
+
+	if o.PrivateKey != nil {
+		n.PrivateKey = o.PrivateKey
+	}
+
+	// reset unmerged leaves since the node is being updated.
+	n.UnmergedLeaves = []leafIndex{}
+}
+
+func (n *RatchetTreeNode) AddUnmerged(l leafIndex) {
+	n.UnmergedLeaves = append(n.UnmergedLeaves, l)
+}
+
+//// Ratchet Tree
+type RatchetTree struct {
```

Comment:
The types in this file should probably all be private.  They're internal to the protocol.  But we can leave them public for now and suppress them later.